src/hotspot/share/gc/g1/g1FullGCCompactTask.cpp

```diff
@@ -35,18 +36,25 @@
 #include "utilities/ticks.hpp"
 
 class G1ResetPinnedClosure : public HeapRegionClosure {
-  G1CMBitMap* _bitmap;
+  G1FullCollector* _collector;
 
 public:
-  G1ResetPinnedClosure(G1CMBitMap* bitmap) : _bitmap(bitmap) { }
+  G1ResetPinnedClosure(G1FullCollector* collector) : _collector(collector) { }
 
   bool do_heap_region(HeapRegion* r) {
-    if (!r->is_pinned()) {
+    uint hr_index = r->hrm_index();
+    // In the prepare phase, we "pin" the regions with high survival ratio
+    // by _region_attr_table, so here we use _region_attr_table rather than
+    // HeapRegion itself to tell whether a region is pinned.
```

Comment:
This comment is going to be completely incomprehensible after a week. While it points out a particular case of pinned regions, this comment is misplaced and immediately begs the question about other cases.

Maybe in the `G1FullGCHeapRegionAttr` class documentation, point out that in particular, that data structure collects whether a region should be considered pinned during full gc (only), and that there are two reasons a region is pinned (and excluded from compaction):
- the HeapRegion itself has been pinned at the start of Full GC
- the occupancy of the region is too high to be considered eligible for compaction